core/src/test/scala/unit/kafka/server/AbstractApiVersionsRequestTest.scala

```diff
@@ -59,15 +68,40 @@ abstract class AbstractApiVersionsRequestTest(cluster: ClusterInstance) {
     } finally socket.close()
   }
 
-  def validateApiVersionsResponse(apiVersionsResponse: ApiVersionsResponse): Unit = {
-    val expectedApis = ApiKeys.zkBrokerApis()
+  def validateApiVersionsResponse(apiVersionsResponse: ApiVersionsResponse, listenerName: ListenerName = cluster.clientListener()): Unit = {
+    val expectedApis = if (!cluster.isKRaftTest) {
+      ApiKeys.zkBrokerApis()
+    } else if (cluster.controllerListenerName().asScala.contains(listenerName)) {
+      ApiKeys.controllerApis()
+    } else {
+      val apis = ApiVersionsResponse.intersectForwardableApis(
+        ApiMessageType.ListenerType.BROKER,
+        RecordVersion.current,
+        new NodeApiVersions(ApiKeys.controllerApis().asScala.map(ApiVersionsResponse.toApiVersion).asJava).allSupportedApiVersions()
+      )
+      apis.add(
+        new ApiVersionsResponseData.ApiVersion()
```

Comment:
This had me a bit confused. It looks like the kraft broker is exposing `Envelope` request because of the logic here: https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/ApiVersionManager.scala#L104. This seems like a bug to me since the kraft broker itself should not receive `Envelope` requests. I feel like we should fix this so that we only add `Envelope` request to `ApiVersions` if forwarding is enabled _and_ it is not a kraft broker.

Reply:
I made a more radical change here to directly remove Envelope from the `ApiVersions`, so we will not support envelop request until we change the `EnvelopeRequest.json` to add zkbroker to its listeners.